src/hardware/opl.cpp

```diff
@@ -488,46 +492,92 @@ void OPL::Init(const uint16_t sample_rate)
 	}
 }
 
-void OPL::WriteReg(const uint32_t reg, const uint8_t val)
+void OPL::WriteReg(const io_port_t selected_reg, const uint8_t val)
 {
-	OPL3_WriteRegBuffered(&oplchip, static_cast<uint16_t>(reg), val);
-	if (reg == 0x105)
-		newm = reg & 0x01;
+	OPL3_WriteRegBuffered(&oplchip, selected_reg, val);
+	if (selected_reg == 0x105)
+		newm = selected_reg & 0x01;
 }
 
-uint32_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
+io_port_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
 {
-	uint16_t addr;
-	addr = val;
+	io_port_t addr = val;
 	if ((port & 2) && (addr == 0x05 || newm)) {
 		addr |= 0x100;
 	}
 	return addr;
 }
 
-void OPL::Generate(const mixer_channel_t &chan, const uint16_t frames)
+AudioFrame OPL::RenderFrame()
```

Comment:
Much nicer interface 👍🏻

Reply:
Your additions to the `AudioFrame` make it a really nice fit!